File: native-wrapper/ios-native/GnhNotifications/GnhNotificationsModule.swift
import Foundation
import React
import UserNotifications

/** RN bridge for GnhNotifications — payloads are JSON strings from the WebView bridge. */
@objc(GnhNotifications)
class GnhNotificationsModule: RCTEventEmitter {
  private let publisher = GnhNotificationPublisher()
  private var hasListeners = false

  override static func requiresMainQueueSetup() -> Bool { false }

  override func supportedEvents() -> [String]! {
    ["gnhNotificationTap"]
  }

  override func startObserving() {
    hasListeners = true
    GnhNotificationCenterDelegate.shared.install()
    GnhNotificationCenterDelegate.shared.onNotificationTap = { [weak self] eventId in
      guard let self, self.hasListeners else { return }
      self.sendEvent(withName: "gnhNotificationTap", body: eventId)
    }
  }

  override func stopObserving() {
    hasListeners = false
  }

  @objc func applyPrivacySettings(
    _ settingsJson: String,
    resolver: @escaping RCTPromiseResolveBlock,
    rejecter: RCTPromiseRejectBlock
  ) {
    guard let settings = parseJson(settingsJson) else {
      resolver(false)
      return
    }
    let notificationsEnabled = settings["notificationsEnabled"] as? Bool ?? false
    if !notificationsEnabled {
      publisher.cancelAllFeatureNotifications()
    }
    resolver(true)
  }

  @objc func publishEvent(
    _ payloadJson: String,
    resolver: @escaping RCTPromiseResolveBlock,
    rejecter: RCTPromiseRejectBlock
  ) {
    guard let payload = parseJson(payloadJson),
          let eventId = payload["eventId"] as? String else {
      resolver(GnhPublishOutcome.disabled.rawValue)
      return
    }
    let input = GnhPublishEventInput(
      eventId: eventId,
      title: payload["title"] as? String ?? "Get NowHere",
      body: payload["body"] as? String ?? "New message",
      badgeCount: payload["badgeCount"] as? Int ?? 0,
      notificationsEnabled: payload["notificationsEnabled"] as? Bool ?? false,
      bannersEnabled: payload["bannersEnabled"] as? Bool ?? false,
      appInForeground: payload["appInForeground"] as? Bool ?? true
    )
    publisher.publishEvent(input) { outcome in
      resolver(outcome.rawValue)
    }
  }

  @objc func setBadgeCount(
    _ count: NSNumber,
    resolver: @escaping RCTPromiseResolveBlock,
    rejecter: RCTPromiseRejectBlock
  ) {
    publisher.setBadgeCount(count.intValue)
    resolver(true)
  }

  @objc func clearBadge(
    _ resolver: @escaping RCTPromiseResolveBlock,
    rejecter: RCTPromiseRejectBlock
  ) {
    publisher.clearBadge()
    resolver(true)
  }

  @objc func getPermissionStatus(
    _ resolver: @escaping RCTPromiseResolveBlock,
    rejecter: RCTPromiseRejectBlock
  ) {
    UNUserNotificationCenter.current().getNotificationSettings { settings in
      let status: String
      switch settings.authorizationStatus {
      case .authorized: status = "authorized"
      case .provisional: status = "provisional"
      case .denied: status = "denied"
      case .notDetermined: status = "notDetermined"
      default: status = "unknown"
      }
      let result: [String: Any] = [
        "status": status,
        "alert": settings.alertSetting == .enabled,
        "badge": settings.badgeSetting == .enabled,
      ]
      resolver(self.jsonString(result))
    }
  }

  /** Must run from a foreground user gesture (Settings toggle flow). */
  @objc func requestPermissions(
    _ badge: Bool,
    alert: Bool,
    resolver: @escaping RCTPromiseResolveBlock,
    rejecter: RCTPromiseRejectBlock
  ) {
    publisher.requestPermissions(badge: badge, alert: alert) { granted, status in
      let statusName: String
      switch status {
      case .authorized: statusName = "authorized"
      case .provisional: statusName = "provisional"
      case .denied: statusName = "denied"
      case .notDetermined: statusName = "notDetermined"
      default: statusName = "unknown"
      }
      resolver(self.jsonString(["granted": granted, "status": statusName]))
    }
  }

  private func parseJson(_ raw: String) -> [String: Any]? {
    guard let data = raw.data(using: .utf8) else { return nil }
    return (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
  }

  private func jsonString(_ dict: [String: Any]) -> String {
    guard let data = try? JSONSerialization.data(withJSONObject: dict),
          let str = String(data: data, encoding: .utf8) else {
      return "{}"
    }
    return str
  }
}

File: native-wrapper/android-native/GnhNotifications/GnhNotificationsModule.kt
package im.getnowhere.app.notifications

import android.Manifest
import android.os.Build
import androidx.core.app.ActivityCompat
import androidx.core.app.NotificationManagerCompat
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import org.json.JSONObject

/** RN bridge for GnhNotifications — payloads are JSON strings from the WebView bridge. */
class GnhNotificationsModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    private val publisher by lazy { GnhNotificationPublisher(reactApplicationContext) }

    override fun getName(): String = "GnhNotifications"

    @ReactMethod
    fun applyPrivacySettings(settingsJson: String, promise: Promise) {
        try {
            val settings = JSONObject(settingsJson)
            val notificationsEnabled = settings.optBoolean("notificationsEnabled", false)
            if (notificationsEnabled) {
                publisher.ensureChannel()
            } else {
                // Notifications turned off: remove any banners this feature posted.
                publisher.cancelAll()
            }
            promise.resolve(true)
        } catch (e: Exception) {
            promise.reject("ERR_APPLY_SETTINGS", e.message)
        }
    }

    @ReactMethod
    fun publishEvent(payloadJson: String, promise: Promise) {
        try {
            val payload = JSONObject(payloadJson)
            val outcome =
                publisher.publishEvent(
                    PublishEventInput(
                        eventId = payload.getString("eventId"),
                        title = payload.optString("title", "Get NowHere"),
                        body = payload.optString("body", "New message"),
                        badgeCount = payload.optInt("badgeCount", 0),
                        notificationsEnabled =
                            payload.optBoolean("notificationsEnabled", false),
                        bannersEnabled = payload.optBoolean("bannersEnabled", false),
                        appInForeground = payload.optBoolean("appInForeground", true),
                    ),
                )
            promise.resolve(outcome.name)
        } catch (e: Exception) {
            promise.reject("ERR_PUBLISH", e.message)
        }
    }

    @ReactMethod
    fun setBadgeCount(count: Int, promise: Promise) {
        // Android has no public app-badge API without a notification; launcher
        // badge support varies by OEM. The channel badge + setNumber cover it.
        promise.resolve(true)
    }

    @ReactMethod
    fun clearBadge(promise: Promise) {
        publisher.cancelAll()
        promise.resolve(true)
    }

    @ReactMethod
    fun getPermissionStatus(promise: Promise) {
        try {
            val enabled =
                NotificationManagerCompat.from(reactApplicationContext)
                    .areNotificationsEnabled()
            val result =
                JSONObject()
                    .put("status", if (enabled) "authorized" else "denied")
                    .put("alert", enabled)
                    .put("badge", enabled)
            promise.resolve(result.toString())
        } catch (e: Exception) {
            promise.reject("ERR_STATUS", e.message)
        }
    }

    /** Must run from a foreground user gesture (Settings toggle flow). */
    @ReactMethod
    fun requestPermissions(badge: Boolean, alert: Boolean, promise: Promise) {
        try {
            publisher.ensureChannel()
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) {
                promise.resolve(
                    JSONObject()
                        .put("granted", publisher.hasPostPermission())
                        .put("status", "authorized")
                        .toString(),
                )
                return
            }
            if (publisher.hasPostPermission()) {
                promise.resolve(
                    JSONObject().put("granted", true).put("status", "authorized").toString(),
                )
                return
            }
            val activity = reactApplicationContext.currentActivity
            if (activity == null) {
                promise.resolve(
                    JSONObject().put("granted", false).put("status", "unavailable").toString(),
                )
                return
            }
            ActivityCompat.requestPermissions(
                activity,
                arrayOf(Manifest.permission.POST_NOTIFICATIONS),
                POST_NOTIFICATIONS_REQUEST_CODE,
            )
            // Async result lands in onRequestPermissionsResult; report requested state.
            promise.resolve(
                JSONObject().put("granted", false).put("status", "requested").toString(),
            )
        } catch (e: Exception) {
            promise.reject("ERR_REQUEST_PERMISSIONS", e.message)
        }
    }

    companion object {
        const val POST_NOTIFICATIONS_REQUEST_CODE = 7301
    }
}
